tests/sentry/relay/config/test_metric_extraction.py

```diff
@@ -1160,7 +1226,44 @@ def test_get_metrics_extraction_config_features_combinations(
             assert config is None
         else:
             assert config is not None
-            assert len(config["metrics"]) == number_of_metrics
+            assert len(config["metrics"]) == number_of_metrics * len(
+                OnDemandMetricSpecVersioning.spec_versions
+            )
+
+
+def _metric_spec(tags: list[dict[str, str]]) -> dict[str, Any]:
+    return {
+        "category": "transaction",
+        "condition": {"name": "event.duration", "op": "gte", "value": 10.0},
+        "field": None,
+        "mri": "c:transactions/on_demand@none",
+        "tags": tags,
+    }
+
+
+@django_db_all
+def test_include_environment_for_widgets(default_project: Project) -> None:
+    env_tag = {"field": "event.environment", "key": "environment"}
+    query = "transaction.duration:>=10"
+
+    with Feature([ON_DEMAND_METRICS, ON_DEMAND_METRICS_WIDGETS]):
+        create_widget(["count()"], query, default_project)
```

Comment:
@k-fish should this create_widget call include `{"field": "event.environment", "key": "environment"}` in the groupbys? (In other words, in the query hash)

What about if the create_widget call included columns with the environment?
Like this: `create_widget(["count()"], query, default_project, columns=["environment"])`?

Reply:
> should this create_widget call include {"field": "event.environment", "key": "environment"} in the groupbys? (In other words, in the query hash)

Yes, every widget should always have `environment` in the group-by, which turns into `event.environment` in the metric spec. That saves environment as a tag until the metrics table, so when we query it later we can switch environment on the fly if a user changes env on their dashboard.

`create_widget(["count()"], query, default_project, columns=["environment"])`

Yeah the `columns` w/ `environment` means someone made a widget that explicitly grouped on environment, so we wouldn't repeat the tag.